File: frontend/src/services/audit.ts
import apiClient from '../lib/apiClient';

export interface AuditLog {
  id: number;
  user_id: number | null;
  action: string;
  resource_type: string;
  resource_id: number | null;
  changes: Record<string, any> | null;
  ip_address: string | null;
  user_agent: string | null;
  metadata: Record<string, any> | null;
  created_at: string;
}

export interface AuditLogList {
  items: AuditLog[];
  total: number;
  skip: number;
  limit: number;
}

export interface AuditLogFilters {
  user_id?: number;
  action?: string;
  resource_type?: string;
  resource_id?: number;
  start_date?: string;
  end_date?: string;
}

export const auditService = {
  async getAuditLogs(
    skip: number = 0,
    limit: number = 100,
    filters?: AuditLogFilters
  ): Promise<AuditLogList> {
    const params = new URLSearchParams({
      skip: skip.toString(),
      limit: limit.toString(),
    });

    if (filters?.user_id) params.append('user_id', filters.user_id.toString());
    if (filters?.action) params.append('action', filters.action);
    if (filters?.resource_type) params.append('resource_type', filters.resource_type);
    if (filters?.resource_id) params.append('resource_id', filters.resource_id.toString());

    const { data } = await apiClient.get(`/api/v1/audit?${params}`);
    return data;
  },

  async getMyAuditLogs(
    skip: number = 0,
    limit: number = 100,
    filters?: Omit<AuditLogFilters, 'user_id'>
  ): Promise<AuditLogList> {
    const params = new URLSearchParams({
      skip: skip.toString(),
      limit: limit.toString(),
    });

    if (filters?.action) params.append('action', filters.action);
    if (filters?.resource_type) params.append('resource_type', filters.resource_type);

    const { data } = await apiClient.get(`/api/v1/audit/me?${params}`);
    return data;
  },
};

File: backend/app/api/v1/endpoints/audit.py
"""
Audit log endpoints.
"""
from typing import Any, Optional
from fastapi import APIRouter, Depends, Query
from sqlalchemy.orm import Session

from app.core.deps import get_current_active_user, get_current_superuser
from app.db.database import get_db
from app.models.user import User
from app.services.audit_service import AuditService
from app.schemas.audit_log import AuditLogList, AuditLogFilter

router = APIRouter()


@router.get("/", response_model=AuditLogList)
def get_audit_logs(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    user_id: Optional[int] = None,
    action: Optional[str] = None,
    resource_type: Optional[str] = None,
    resource_id: Optional[int] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_superuser)
) -> Any:
    """Get audit logs (Admin only)."""
    filters = AuditLogFilter(
        user_id=user_id,
        action=action,
        resource_type=resource_type,
        resource_id=resource_id
    )
    
    logs = AuditService.get_logs(db, filters, skip, limit)
    total = AuditService.count_logs(db, filters)
    
    return {
        "items": logs,
        "total": total,
        "skip": skip,
        "limit": limit
    }


@router.get("/me", response_model=AuditLogList)
def get_my_audit_logs(
    skip: int = Query(0, ge=0),
    limit: int = Query(100, ge=1, le=1000),
    action: Optional[str] = None,
    resource_type: Optional[str] = None,
    db: Session = Depends(get_db),
    current_user: User = Depends(get_current_active_user)
) -> Any:
    """Get current user's audit logs."""
    filters = AuditLogFilter(
        user_id=current_user.id,
        action=action,
        resource_type=resource_type
    )
    
    logs = AuditService.get_logs(db, filters, skip, limit)
    total = AuditService.count_logs(db, filters)
    
    return {
        "items": logs,
        "total": total,
        "skip": skip,
        "limit": limit
    }
